src/Glpi/Asset/CustomField.php

```diff
@@ -0,0 +1,363 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+namespace Glpi\Asset;
+
+use CommonDBChild;
+use CommonGLPI;
+use DirectoryIterator;
+use Dropdown;
+use Glpi\Application\View\TemplateRenderer;
+use Glpi\Asset\CustomField\TypeInterface;
+use Glpi\DBAL\QueryExpression;
+use Glpi\DBAL\QueryFunction;
+use ReflectionClass;
+use Session;
+
+final class CustomField extends CommonDBChild
+{
+    public static $itemtype = AssetDefinition::class;
+    public static $items_id = 'assets_assetdefinitions_id';
+
+    public static $rightname = 'config';
+
+    public static function getTypeName($nb = 0)
+    {
+        return _n('Custom field', 'Custom fields', $nb);
+    }
+
+    public static function getIcon()
+    {
+        return 'ti ti-forms';
+    }
+
+    public static function canCreate(): bool
+    {
+        return parent::canUpdate();
+    }
+
+    public static function canDelete(): bool
+    {
+        return parent::canUpdate();
+    }
+
+    public static function canPurge(): bool
+    {
+        return parent::canUpdate();
+    }
+
+    public function cleanDBonPurge()
+    {
+        /** @var \DBmysql $DB */
+        global $DB;
+
+        $DB->update('glpi_assets_assets', [
+            'custom_fields' => QueryFunction::jsonRemove([
+                'custom_fields',
+                new QueryExpression($DB::quoteValue('$."' . $this->fields['id'] . '"'))
+            ])
+        ], [
+            'assets_assetdefinitions_id' => $this->fields['assets_assetdefinitions_id'],
+        ]);
+    }
+
+    public static function getAllowedDropdownItemtypes()
+    {
+        /**
+         * @var array $CFG_GLPI
+         */
+        global $CFG_GLPI;
+
+        static $allowed_dropdown_itemtypes = null;
+
+        if ($allowed_dropdown_itemtypes === null) {
+            $allowed_dropdown_itemtypes = [
+                _n('Asset', "Assets", Session::getPluralNumber()) => array_combine(
+                    $CFG_GLPI['asset_types'],
+                    array_map(static fn ($t) => $t::getTypeName(1), $CFG_GLPI['asset_types'])
+                ),
+            ];
+            $allowed_dropdown_itemtypes = array_merge_recursive($allowed_dropdown_itemtypes, Dropdown::getStandardDropdownItemTypes());
+        }
+
+        return $allowed_dropdown_itemtypes;
+    }
+
+    /**
+     * Get the list of field types available
+     * @return array<string, class-string<TypeInterface>>
+     */
+    public static function getFieldTypes(): array
+    {
+        // Get files in the current directory
+        $directory_iterator = new DirectoryIterator(__DIR__ . '/CustomField');
+        static $all_types = null;
```

Comment:
Please do not use `static` variables.

This could probably be moved into the `AssetDefinitionManager` constructor, as we do for the capacities. It would be easier to handle in the testing context.